models/issue.go

```diff
@@ -1909,3 +1881,129 @@ func (issue *Issue) updateClosedNum(e Engine) (err error) {
 	}
 	return
 }
+
+// ResolveMentionsByVisibility returns the users mentioned in an issue, removing those that
+// don't have access to reading it. Teams are expanded into their users, but organizations are ignored.
+func (issue *Issue) ResolveMentionsByVisibility(ctx DBContext, doer *User, mentions []string) (users []*User, err error) {
+	if len(mentions) == 0 {
+		return
+	}
+	if err = issue.loadRepo(ctx.e); err != nil {
+		return
+	}
+	resolved := make(map[string]bool, len(mentions))
+	for _, name := range mentions {
+		name := strings.ToLower(name)
+		if _, ok := resolved[name]; ok {
+			continue
+		}
+		resolved[name] = (name == doer.LowerName)
+	}
```

Comment:
Why not just `append(names, name)` here?

Reply:
To avoid including the same user twice. For example:
```
>
>
> This quote was already mentioning @mike 

This message is for @mike and @thewholeteam
```

Reply:
@lunny To avoid including the same user twice. For example:
```
>
>
> This quote was already mentioning @mike 

This message is for @mike and @thewholeteam
```

Reply:
I mean
```go
resolved := make(map[string]bool, len(mentions))
names := make([]string, 0, len(resolved))
for _, name := range mentions {
		name := strings.ToLower(name)
		if _, ok := resolved[name]; ok {
			continue
		}
		resolved[name] = (name == doer.LowerName)
               if name != doer.LowerName {
                      names = append(names, name)
               }
	}
```

Reply:
@lunny done. I've also found that I was inadvertently including the doer when processing teams... and had it cleared in the test! :flushed: